### PR TITLE
Opentracing middleware

### PR DESCRIPTION
**Pull request by bodhi**
Introduce tracing middleware to trace HTTP requests and push traces into Jaeger.

### CODE REVIEW

**Comment by bodhi** [comment]
@junhuif @lifenod this adds middleware and support for tracing requests. You need to add this to the server configuration:

```go
	closer, tracer, err := tracing.Tracer(log)
	if err != nil {
		log.WithError(err).Log()
	} else {
		defer closer.Close()
	}

	// ...
		server.Compose(
			// other middleware...
			tracer,
			server.DefaultMiddleware(log),
		)
```

The tracing middleware should be near the bottom of the middleware stack. At least it should be *below* `errornotifier` and `montioring`. Because `errornotifier` and `monitoring` will trace API requests to Airbrake and InfluxDB.

**Comment by bodhi** [comment]
The tracing client is configured by environment variables: https://github.com/jaegertracing/jaeger-client-node#environment-variables

I'll work with @theplant/sa to make Jaeger available for recording traces.

File: tracing/README.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+   ```
+   http.ListenAndServe(tracer(handler))
+   ```
+
+   If you're using `server.Compose`, add the tracer to the bottom of
+   your stack, just above `server.DefaultMiddleware`. This will make
+   the tracing system available to other middleware such as
+   `errornotifier` (see below).
+
+# Trace Propagagtion
```
Comment (bodhi) [inline tracing/README.md]:
```diff
-# Trace Propagagtion
+# Trace Propagation
```

[the remark this answers is not in the document]
Comment (bodhi) [reply tracing/README.md]: Fixed in 55c4039

File: tracing/tracing.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+	e = f(ctx, span)
+	return
+}
+
+type nullCloser struct{}
+
+func (nullCloser) Close() error { return nil }
+
+// Tracer is used to create tracing middleware that uses jager
```
[the remark this answers is not in the document]
Comment (bodhi) [reply tracing/tracing.go]: Fixed in 55c4039

[the remark this answers is not in the document]
Comment (bodhi) [reply tracing/tracing.go]: Fixed in eec3742

File: tracing/tracing.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+func trace(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		l := log.ForceContext(ctx).With(
+			"context", "appkit/tracing.trace",
+		)
+
+		// Extract tracing propagarion info from HTTP request
```
Comment (lifenod) [inline tracing/tracing.go]: the same typo?
propagarion -> propagation

**Comment by lifenod** [comment]
## [OT]

Do we plan to add tracing to the server?
For example, route, load balancing, k8s.
I think it's useful to analyze performance problems except for app.

**Comment by idealhack** [comment]
We might need service mesh (e.g., istio) for that... I think we'll know it when we need it.

File: tracing/tracing.go (hunk 3 above)
Comment (junhuif) [reply tracing/tracing.go]: Fixes in 6a0681bd0a86cf9028b6129f34441cda1ac84ff2.

**Comment by bodhi** [comment]
> Do we plan to add tracing to the server side?

https://github.com/opentracing-contrib/nginx-opentracing looks promising. I think though that we are *a long way* from needing a service mesh, as @idealhack said.

**Pull request merged**